feat(js): Add SDK Options page (#12502)
getsentry/sentry-docs: docs/platforms/javascript/common/options.mdx

---
title: SDK Options
description: "Learn more about how to configure the SDK. These options are set when the SDK is first initialized, passed to the init function as an object."
customCanonicalTag: "/platforms/javascript/options/"
sidebar_order: 2
---

<PlatformContent includePath="configuration/config-intro" />

## Available Options

<SdkOptionsOverview />

## Core Options

<SdkOption name='dsn' type='string' envVar='SENTRY_DSN'>

> The _DSN_ tells the SDK where to send the events. If this value is not provided, the SDK will try to read it from the `SENTRY_DSN` environment variable. If that variable also does not exist, the SDK will just not send any events.

Reviewer: since we talked about env vars in the morning: We should change this description for client vs server

In runtimes without a process environment (such as the browser) that fallback does not apply.

Learn more about [DSN utilization](/product/sentry-basics/dsn-explainer/#dsn-utilization).

</SdkOption>

<SdkOption name="debug" type='boolean' defaultValue='false'>

> Turns debug mode on or off. If debug is enabled SDK will attempt to print out useful debugging information if something goes wrong with sending the event. The default is always `false`. It's generally not recommended to turn it on in production, though turning `debug` mode on will not cause any safety concerns.

Reviewer: safety sounds weird here 🤔 either we say "security" or write something like it won't expose any critical information? Or just omit it?

```suggestion
Turns debug mode on or off. If debug is enabled SDK will attempt to print out useful debugging information if something goes wrong with sending the event. The default is always `false`. It's generally not recommended to turn it on in production, though turning `debug` mode on will not cause any security concerns.
```

Author: maybe we should just remove the last sentence here, not sure how relevant this is actually..?

</SdkOption>

<SdkOption name="sampleRate" type='number' defaultValue='1.0'>

Configures the sample rate for error events, in the range of `0.0` to `1.0`. The default is `1.0`, which means that 100% of error events will be sent. If set to `0.1`, only 10% of error events will be sent. Events are picked randomly.

</SdkOption>

<SdkOption name="maxBreadcrumbs" type='number' defaultValue="100">

This variable controls the total amount of breadcrumbs that should be captured. You should be aware that Sentry has a [maximum payload size](https://develop.sentry.dev/sdk/data-model/envelopes/#size-limits) and any events exceeding that payload size will be dropped.

</SdkOption>

<SdkOption name="attachStacktrace" type='boolean' defaultValue='false'>

> When enabled, stack traces are automatically attached to all messages logged. Stack traces are always attached to exceptions; however, when this option is set, stack traces are also sent with messages. This option, for instance, means that stack traces appear next to all log messages.

Reviewer: l: maybe this makes it a bit clearer?

```suggestion
When enabled, stack traces are automatically attached to all messages logged. Stack traces are always attached to exceptions; however, when this option is set, stack traces are also sent with messages. This option, for instance, means that stack traces appear next to all messages captured with `Sentry.captureMessage()`.
```

<Include name="platforms/configuration/options/deny-urls.mdx" />

</SdkOption>

<SdkOption name="allowUrls" type='Array<string | RegExp>' defaultValue='[]'  categorySupported={['browser']}>

<Include name="platforms/configuration/options/allow-urls.mdx" />

</SdkOption>

<PlatformSection supported={['javascript.electron']}>
<SdkOption name="autoSessionTracking" type='boolean' defaultValue='true'>

When not set to `false`, the SDK tracks sessions linked to the lifetime of the Electron main process.

</SdkOption>
</PlatformSection>

<SdkOption name="initialScope" type="CaptureContext">

Data to be set to the initial scope. Initial scope can be defined either as an object or a callback function, as shown below.

```javascript
// Using an object
Sentry.init({
  dsn: "___PUBLIC_DSN___",
  initialScope: {
    tags: { "my-tag": "my value" },
    user: { id: 42, email: "john.doe@example.com" },
  },
});

// Using a callback function
Sentry.init({
  dsn: "___PUBLIC_DSN___",
  initialScope: (scope) => {
    scope.setTags({ a: "b" });
    return scope;
  },
});
```

</SdkOption>

<SdkOption name="maxValueLength" type='number' defaultValue='250'>

> Maximum number of characters a single value can have before it will be truncated.

Reviewer: l: probably a bit clearer? (feel free to disregard)

```suggestion
Maximum number of characters a every string property on events sent to Sentry can have before it will be truncated.
```

</SdkOption>

<SdkOption name="enabled" type='boolean' defaultValue='true'>

Specifies whether this SDK should send events to Sentry. Setting this to `enabled: false` doesn't prevent all overhead from Sentry instrumentation. To disable Sentry completely, depending on environment, call `Sentry.init` conditionally.

</SdkOption>

<SdkOption name="sendClientReports" type='boolean' defaultValue='true'>

Set this option to `false` to disable sending of client reports. Client reports are a protocol feature that let clients send status reports about themselves to Sentry. They are currently mainly used to emit outcomes for events that were never sent.

</SdkOption>

<SdkOption name="includeLocalVariables" type='boolean' defaultValue='false'  categorySupported={['server', 'serverless']}>

Set this option to `true` to add stack local variables to stack traces.

For more advanced configuration options, see the documentation on the <PlatformLink to="/configuration/integrations/localvariables">Local Variables integration options</PlatformLink>.

</SdkOption>

> <SdkOption name="integrations" type='Array<Integration>' defaultValue='[]'>

Reviewer: ```suggestion
<SdkOption name="integrations" type='Array<Integration> | (integrations: Array<Integration>) => Array<Integration>' defaultValue='[]'>
```

Pass additional integrations that should be initialized with the SDK. Integrations are pieces of code that can be used to extend the SDK's functionality. They can be used to add custom event processors, context providers, or to hook into the SDK's lifecycle.

See <PlatformLink to="/configuration/integrations/">integration docs</PlatformLink> for more information.

</SdkOption>

<SdkOption name="defaultIntegrations" type='undefined | false'>

This can be used to disable integrations that are added by default. When set to `false`, no default integrations are added.

See <PlatformLink to="/configuration/integrations/#modifying-default-integrations">integration docs</PlatformLink> to see how you can modify the default integrations.

</SdkOption>

<SdkOption name="transport" type="(transportOptions: TransportOptions) => Transport">

The JavaScript SDK uses a transport to send events to Sentry. On modern browsers, most transports use the browsers' fetch API to send events. Transports will drop an event if it fails to send due to a lack of connection.

<PlatformCategorySection supported={["browser"]}>

>   In the browser, a fetch-based transport is used by default.

Reviewer: l:
```suggestion
  In the browser, a `fetch`-based transport is used by default.
```

</PlatformCategorySection>

<PlatformCategorySection supported={["server", "serverless"]}>
  On the server, a `https`-based transport is used by default.
</PlatformCategorySection>

</SdkOption>

<SdkOption name="transportOptions" type='TransportOptions'>

Options used to configure the transport. This is an object with the following possible optional keys:

<PlatformCategorySection supported={["server", "serverless"]}>
  - `headers`: An object containing headers to be sent with every request. -
  `proxy`: A proxy used for outbound requests. Can be http or https. -
  `caCerts`: A path or list of paths to a CA certificate, or a buffer of CA
  certificates. - `httpModule`: A custom HTTP module to use for requests.
  Defaults to the the native `http` and `https` modules. - `keepAlive`:
  Determines whether to keep the socket alive between requests. Defaults to
  `false`.
</PlatformCategorySection>

<PlatformCategorySection supported={["browser", "desktop"]}>
  - `headers`: An object containing headers to be sent with every request. Used

>   by the SDK's fetch and XHR transports. - `fetchOptions`: An object containing

Reviewer: l:
```suggestion
  by the SDK's `fetch` and XHR transports. - `fetchOptions`: An object containing
```

>   by the SDK's fetch and XHR transports. - `fetchOptions`: An object containing

Reviewer: are we intentionally not putting the individual transport options into their own lines?

Author: no! auto-format messed this up, good catch!

</SdkOption>

<SdkOption name="tracesSampler" type='(samplingContext: SamplingContext) => number | boolean'>

A function responsible for determining the percentage chance a given transaction will be sent to Sentry. It will automatically be passed information about the transaction and the context in which it's being created, and must return a number between `0` (0% chance of being sent) and `1` (100% chance of being sent). Can also be used for filtering transactions, by returning 0 for those that are unwanted. Either this or `tracesSampleRate` must be defined to enable tracing.

The `samplingContext` object passed to the function has the following properties:

- `parentSampled`: The sampling decision of the parent transaction. This is `true` if the parent transaction was sampled, and `false` if it was not.
- `name`: The name of the span as it was started.
- `attributes`: The initial attributes of the span.
  <PlatformCategorySection supported={["server", "serverless"]}>
    - `normalizedRequest`: Request information of the currently active HTTP
    server request, if applicable.
  </PlatformCategorySection>

</SdkOption>

> <SdkOption name="tracePropagationTargets" type='Array<string | RegExp>'>

Reviewer: I'd also add a default value entry into the table. Which needs to be different on browser and server.

Author: yeah this is a bit tricky because 1. it is hard to implement this nicely (we can, if we want to) and especially 2. the default value here is not an actual value but computed (e.g. based on the current host), which is why I opted to rather explain the default in this case?

Reviewer: idea (logaf-l) should we then add something liks "see description" for the default value field?
